### PR TITLE
ENH: Modify scipy.optimize.least_squares to accept bounds of type Bounds

### PR DESCRIPTION
**Pull request by FranzForstmayr**
#### What does this implement/fix?
Modified `least_squares` function to accept instances of type `Bounds` for the `bounds` parameter.

I wanted to interchange the functions `optimize.minimize` and `optimize.least_squares` easily, but unfortunately the array shape needed for bounds is transposed for these two functions.
`optimize.minimize` also accepts instances of `Bounds`, so I thought this would be a quick fix to use both methods with the same `Bounds` variable.

### CODE REVIEW

File: scipy/optimize/_lsq/least_squares.py [hunk 1]
```diff
@@ -4,10 +4,11 @@
 import numpy as np
 from numpy.linalg import norm
 
-from scipy.sparse import issparse, csr_matrix
```
Comment (AtsushiSakai) [inline scipy/optimize/_lsq/least_squares.py:7]: I'm not sure why `csr_matrix` import is needed in this PR?

File: scipy/optimize/_lsq/least_squares.py [hunk 2]
```diff
@@ -793,7 +792,17 @@ def least_squares(
...
+    if isinstance(bounds, Bounds):
+        lb, ub = bounds.lb, bounds.ub
+        bounds = (lb, ub)
+    else:
+        if len(bounds) == 2:
+            lb, ub = prepare_bounds(bounds, x0.shape[0])
+        else:
+            raise ValueError("`bounds` must contain 2 elements.")
+
+    if max_nfev is not None and max_nfev <= 0:
```
Comment (AtsushiSakai) [inline scipy/optimize/_lsq/least_squares.py:804]: Why does the `max_nfev` check move here in this PR?

**Review by AtsushiSakai — COMMENTED** [review]
Thank you for your PR. I think supporting `Bounds` in `least_squeres` is reasonable for me. I have some questions.

File: scipy/optimize/_lsq/least_squares.py (hunk 1 above)
Comment (FranzForstmayr) [reply scipy/optimize/_lsq/least_squares.py:7]: I removed the import of `csr_matrix` as it was unused.

File: scipy/optimize/tests/test_least_squares.py [hunk 3]
```diff
@@ -430,6 +430,27 @@ def test_bounds_shape(self):
                 method=self.method)
             assert_allclose(res.x, [0.0, 0.5], atol=1e-5)
 
+    def test_bound_instances(self):
```
Comment (AtsushiSakai) [inline scipy/optimize/tests/test_least_squares.py:433]:
```suggestion
    def test_bounds_instances(self):
```
Very minor point.

File: scipy/optimize/tests/test_least_squares.py [hunk 4]
```diff
@@ -430,6 +430,27 @@ def test_bounds_shape(self):
...
+        assert_allclose(res.x, -1.0, atol=1e-4)
+
+        res = least_squares(fun_2d_trivial, [0.5, 0.5],
+                            bounds=Bounds(lb=[-1.0, -1.0], ub=1.0))
+        assert_allclose(res.x, [0.0, 0.0], atol=1e-5)
+
+        res = least_squares(fun_2d_trivial, [0.5, 0.5],
+                            bounds=Bounds(lb=[0.1, 0.1]))
+        assert_allclose(res.x, [0.1, 0.1], atol=1e-5)
+
```
Comment (AtsushiSakai) [inline scipy/optimize/tests/test_least_squares.py:453]: Could you please add tests to solve same optimization problems in  `test_bounds_shape` using `Bounds` instance here?
 Especially, we should check it works or not in ragged bounds case like `bounds=([0.3, 0.2], 3.0)` and using `jac` or not.

**Comment by andyfaff** [comment]
@AtsushiSakai, this PR looks in good shape to me. Would you like to merge this now, or are there any outstanding things you'd like to see?

**Comment by AtsushiSakai** [comment]
I just missed this PR is updated. Thank you for letting me know the update.

**Review by AtsushiSakai — APPROVED** [review]
LGTM and another maintainer say this PR looks in good shape, so let's merge it. Thank you for your work. @FranzForstmayr

**Pull request merged**